### PR TITLE
Add default license notice to all Rust files

### PR DESCRIPTION
**Pull request by alevy**
### Pull Request Overview

This pull request adds a default copyright and license notice to all Rust files in the repository

### Testing Strategy

None

### Documentation Updated

- [] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

**Comment by hudson-ayers** [comment]
2 files we may need to consider:
```shell
hudson: ~/tock (master) $ rg Copyright --type rust
capsules/src/sip_hash.rs
16://! Copyright 2012-2016 The Rust Project Developers.
17://! Copyright 2016-2021 Frank Denis.
18://! Copyright 2021 Western Digital

chips/nrf52/src/crt1.rs
11: * Copyright 2016, Michael Andersen <m.andersen@eecs.berkeley.edu>
```

Based on the discussion on the call today, I think our conclusion was that these files should have the added copyright in this PR removed in favor of what is already there. It is also worth noting that the crt1 file seems to be licensed only under Apache 2.0

**Comment by alevy** [comment]
@hudson-ayers thanks for finding those! Fixed those two files (by just removing the lines I had added)

**Comment by jrvanwhy** [comment]
I thought Pat was going to write up a policy document for license headers. Did I misunderstand during the meeting?

**Comment by hudson-ayers** [comment]
> I thought Pat was going to write up a policy document for license headers. Did I misunderstand during the meeting?

You understood correctly, I don't think this will be merged without that document first getting written and approved.

File: arch/cortex-m/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,6 @@
+// Copyright Tock Contributors
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
```
Comment (brghena) [inline arch/cortex-m/src/lib.rs:2]: Should there be parentheses around these? The SPDX website does not show them in its examples except in cases of ambiguity. https://spdx.dev/ids/

**Comment by bradjc** [comment]
Why is the first line Apache or MIT, and the second line MIT or Apache?

**Comment by alevy** [comment]
> Why is the first line Apache or MIT, and the second line MIT or Apache?

Oops... because I made a mistake. Will fix

**Comment by alevy** [comment]
Rebased and added the `Copyright Tock Contributors 2023` line

File: Makefile [hunk 2]
```diff
@@ -1,3 +1,8 @@
+# Licensed under the Apache License, Version 2.0 or the MIT License.
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+# Copyright Tock Contributors 2023
+# Copyright Tock Contributors 2023
```
Comment (jrvanwhy) [inline Makefile:4]: The copyright line should end with a period. Also, it's repeated (I assume a rebase error?).

File: doc/courses/2018-11-SenSys/exercises/app/Makefile [hunk 3]
```diff
@@ -1,3 +1,6 @@
+# Licensed under the Apache License, Version 2.0 or the MIT License.
+# SPDX-License-Identifier: Apache-2.0 OR MIT
```
Comment (jrvanwhy) [inline doc/courses/2018-11-SenSys/exercises/app/Makefile:2]: This file needs a copyright line added.

File: doc/courses/2018-11-SenSys/presentation/Makefile [hunk 4]
```diff
@@ -1,3 +1,6 @@
+# Licensed under the Apache License, Version 2.0 or the MIT License.
+# SPDX-License-Identifier: Apache-2.0 OR MIT
```
Comment (jrvanwhy) [inline doc/courses/2018-11-SenSys/presentation/Makefile:2]: This file needs a copryight notice added.

**Comment by alevy** [comment]
@jrvanwhy ready!

**Review by jrvanwhy — COMMENTED** [review]
I suggest reverting all changes under `tools/license-checker/`. Those files already have license header, so this PR doubles up on the headers. Also, the license checker's unit tests use its own source code's headers, so this change causes the unit tests to fail.

Also, GitHub is reporting merge conflicts that will need to be resolved before merging this PR.

There are some exceptions in `.lcignore` that I don't think belong, but I think we should resolve those in a later PR. I'm opposed to adding more files to this PR because my system is really struggling to run GitHub's UI as-is (due to the number of files).

**Comment by alevy** [comment]
> I suggest reverting all changes under tools/license-checker/.

Done

> Also, GitHub is reporting merge conflicts that will need to be resolved before merging this PR.

Done. Rebased on master

> There are some exceptions in .lcignore that I don't think belong,

I agree. Some of them are because the license checker seems to barf at, e.g., `/* */` style comments that are necessary for linker scripts. Others might be oversights.

**Comment by jrvanwhy** [comment]
> > There are some exceptions in .lcignore that I don't think belong,
> 
> I agree. Some of them are because the license checker seems to barf at, e.g., `/* */` style comments that are necessary for linker scripts. Others might be oversights.

The linker script comment thing is an oversight in the implementation of the license checker. I'll fix that (after this PR though).

**Comment by jrvanwhy** [comment]
There are still some CI errors, but they don't appear to be related to this PR. Is `master` broken right now?

**Comment by jrvanwhy** [comment]
Oh, I think I see the cause of the CI issue. There are headers before the shebang lines in `.sh` and `.py` files. I suggest reverting the changes to `.sh` and `.py` files in this PR; we can deal with them in a later PR.

**Comment by alevy** [comment]
@jrvanwhy done

**Review by hudson-ayers — DISMISSED** [review]

File: boards/components/src/can.rs [hunk 5]
```diff
@@ -1,3 +1,7 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
```
Comment (jrvanwhy) [inline boards/components/src/can.rs:4]: This file already has a copyright header, please revert.

File: capsules/extra/src/can.rs [hunk 6]
```diff
@@ -1,3 +1,7 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
```
Comment (jrvanwhy) [inline capsules/extra/src/can.rs:4]: This file already has a license header, please revert.

**Review by jrvanwhy — COMMENTED** [review]
@lschuermann The following files were added in #3110, should they be copyright 2022 or 2023?
1. [`chips/qemu_rv32_virt_chip/src/virtio_mmio.rs`](https://github.com/tock/tock/tree/master/chips/qemu_rv32_virt_chip/src/virtio_mmio.rs)
2. [`chips/virtio/src/queues/split_queue.rs`](https://github.com/tock/tock/tree/master/chips/virtio/src/queues/split_queue.rs)

Also, the following files were added in #3396, same question (2022 or 2023):
1. [`capsules/core/Cargo.toml`](https://github.com/tock/tock/tree/master/capsules/core/Cargo.toml)
2. [`capsules/core/src/lib.rs`](https://github.com/tock/tock/tree/master/capsules/core/src/lib.rs)
3. [`capsules/core/src/virtualizers/mod.rs`](https://github.com/tock/tock/tree/master/capsules/core/src/virtualizers/mod.rs)
4. [`capsules/extra/Cargo.toml`](https://github.com/tock/tock/tree/master/capsules/extra/Cargo.toml)
5. [`capsules/extra/src/test/mod.rs`](https://github.com/tock/tock/tree/master/capsules/extra/src/test/mod.rs)

**Review by bradjc — DISMISSED** [review]

**Comment by jrvanwhy** [comment]
> @lschuermann The following files were added in #3110, should they be copyright 2022 or 2023?
> 
> 1. [`chips/qemu_rv32_virt_chip/src/virtio_mmio.rs`](https://github.com/tock/tock/tree/master/chips/qemu_rv32_virt_chip/src/virtio_mmio.rs)
> 2. [`chips/virtio/src/queues/split_queue.rs`](https://github.com/tock/tock/tree/master/chips/virtio/src/queues/split_queue.rs)
> 
> Also, the following files were added in #3396, same question (2022 or 2023):
> 
> 1. [`capsules/core/Cargo.toml`](https://github.com/tock/tock/tree/master/capsules/core/Cargo.toml)
> 2. [`capsules/core/src/lib.rs`](https://github.com/tock/tock/tree/master/capsules/core/src/lib.rs)
> 3. [`capsules/core/src/virtualizers/mod.rs`](https://github.com/tock/tock/tree/master/capsules/core/src/virtualizers/mod.rs)
> 4. [`capsules/extra/Cargo.toml`](https://github.com/tock/tock/tree/master/capsules/extra/Cargo.toml)
> 5. [`capsules/extra/src/test/mod.rs`](https://github.com/tock/tock/tree/master/capsules/extra/src/test/mod.rs)

Leon wasn't immediately available, so we decided on the core WG call to change these dates to 2023 so this can be merged ASAP.

**Comment by jrvanwhy** [comment]
You'll need to add `tools/sha256sum/src/main.rs` to a `.lcignore` file for CI to pass.

File: capsules/core/src/virtualizers/mod.rs [hunk 7]
```diff
@@ -1,3 +1,7 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
```
Comment (jrvanwhy) [inline capsules/core/src/virtualizers/mod.rs:3]: It appears this file was created in 2023, not 2022.

**Comment by lschuermann** [comment]
> Leon wasn't immediately available, so we decided on the core WG call to change these dates to 2023 so this can be merged ASAP.

@jrvanwhy My apologies for the delayed response! 2023 is absolutely fine, I can update them right now.

**Comment by lschuermann** [comment]
Okay, raced with @alevy on the push. This looks good to me otherwise, thanks so much for the work!

**Review by lschuermann — APPROVED** [review]

**Review by jrvanwhy — APPROVED** [review]
`capsules/extra/src/can.rs` still has an extra header, but we can fix that in a later PR.

**Pull request merged**